src/viadot/sources/supermetrics.py

```diff
@@ -0,0 +1,266 @@
+        Download query results to a dictionary.
+
+        Parameters
+        ----------
+        timeout : tuple of float, optional
+            The timeout values for the request. Defaults to (3.05, 60 * 30) - a short
+            timeout for connection issues and a long timeout for query execution.
+
+        Returns
+        -------
+        dict
+            The response from the API as a JSON dictionary.
+
+        Raises
+        ------
+        ValueError
+            If query parameters are not set before calling this method.
+        """
+        if not self.query_params:
+            raise ValueError("Please build the query first")
+
+        params = {"json": json.dumps(self.query_params)}
+        headers = {"Authorization": f"Bearer {self.api_key}"}
+
+        response = handle_api_response(
+            url=self.API_ENDPOINT, params=params, headers=headers, timeout=timeout
+        )
+        return response.json()
+
+    @classmethod
+    def _get_col_names_google_analytics(
+        cls,
+        response: dict,
+    ) -> List[str]:
+        """
+        Get column names from Google Analytics data.
+
+        Parameters
+        ----------
+        response : dict
+            Dictionary with the JSON response from the API call.
+
+        Returns
+        -------
+        list of str
+            List of column names for Google Analytics data.
+
+        Raises
+        ------
+        ValueError
+            If no data is returned or if the column names cannot be determined.
+        """
+        is_pivoted = any(
+            field["field_split"] == "column"
+            for field in response["meta"]["query"]["fields"]
+        )
+
+        if is_pivoted:
+            if not response["data"]:
+                raise ValueError(
+                    "Couldn't find column names as query returned no data."
+                )
+            columns = response["data"][0]
+        else:
+            cols_meta = response["meta"]["query"]["fields"]
+            columns = [col_meta["field_name"] for col_meta in cols_meta]
+        return columns
+
+    @classmethod
+    def _get_col_names_other(cls, response: dict) -> List[str]:
+        """
+        Get column names from non-Google Analytics data.
+
+        Parameters
+        ----------
+        response : dict
+            Dictionary with the JSON response from the API call.
+
+        Returns
+        -------
+        list of str
+            List of column names.
+        """
+        cols_meta = response["meta"]["query"]["fields"]
+        columns = [col_meta["field_name"] for col_meta in cols_meta]
+        return columns
+
+    def _get_col_names(self) -> List[str]:
+        """
+        Get column names based on the data type.
+
+        Returns
+        -------
+        list of str
+            List of column names.
+
+        Raises
+        ------
+        ValueError
+            If the column names cannot be determined.
+        """
+        response: dict = self.to_json()
```

Comment:
This seems quite suboptimal - you're calling the API twice, once just to get the columns, then another time to get the data itself. I think extracting the columns and the data should be done with 1 API call.

Reply:
Created an issue to solve this https://github.com/dyvenia/viadot/issues/1023